### ISSUE TITLE
Simplifying Locale handling of extensions

### ISSUE THREAD

**Issue by sffc** [unicode-org/icu4x#1703]
Currently, the Locale class is deeply structured and nested when it comes to extension keywords. This makes it easy to perform processing if you know exactly what you are looking for, but difficult if you want to perform large-scale manipulation of the extensions.

Currently, we have (pseudocode):

```
extensions: {
    unicode: {
        keywords: [(TinyStr, [TinyStr])]
        attributes: [TinyStr]
    }
    transform: {
        lang: LanguageIdentifier
        fields: [(TinyStr, [TinyStr])]
    }
    private: [TinyStr]
    other: [(u8, [TinyStr])]
}
```

Examples of operations I want to perform for which the current structure is poorly suited:

1. Compare a Locale to a LocaleStr (stringified locale).
2. Remove all extension values except for an allowlist.

As far as I am aware, the use cases for which the current design for extensions was chosen are fairly limited. I would therefore like to propose that we take my use cases and consider a design for extensions that optimizes for them.

Here is the structure I would like:

```
extensions: [ExtensionEnum]  // sorted!

enum ExtensionEnum {
    TransformLang(LanguageIdentifier),
    TransformField(TinyStr, [TinyStr]),
    UnicodeAttribute(TinyStr),
    UnicodeKeyword(TinyStr, [TinyStr]),
    Private([TinyStr]),
    Other(u8, [TinyStr]),
}
```

The sort order would be defined to be the same as the proper UTS-35 sort order. This *almost* works out of the box except that we need a custom `Ord` since `Other` could be interleaved.

Needs approval from:

- [x] @zbraniecki 
- [x] @Manishearth

**Comment by zbraniecki** [comment]
I see the value of the current model in that it:
1) Provides much higher level of struct-induced validity. It is impossible to have two `others` or two `private` by accident. You'd need to perform such validation on each operation.
2) It provides highly ergonomic API for developers to work on extensions - retrieve, inspect, filter, modify, add, remove.
3) It allows for semantically valid suboperations - serialize/deserialize unicode extensions, "transplant" transform extension from one locale to another etc.

You make a point that you want to forego this API because we haven't seen extensive use of it. My counter point is that maybe we haven't seen an extensive use of it, because all Locale APIs before ICU4X are really bad at allowing anyone to work with extensions.

> Examples of operations I want to perform for which the current structure is poorly suited:
>
>   1) Compare a Locale to a LocaleStr (stringified locale).
>   2) Remove all extension values except for an allowlist.

I don't fully understand why (1) is hard. The compare function has to navigate a tree, rather than iterate a list, but that's it. It's a single tree, with well defined structure and a high-performance comparison will "zip" compare between those two.

As for (2) shouldn't the allowlist be per-extension as well? It seems that a list of values you'd want to preserve shouldn't be "blank-slate" and should narrow values in particular extensions. Is there a use case which doesn't care?

I'm initially opposing this proposal on the ground of:
1) I don't agree that the use cases you provided are significantly easier with changed model
2) It seems like a loss of semantic consistency for a data structure that is harder to keep valid
3) Canonical operations are becoming more costly

I'm not very strongly opposed, and I can see myself being convinced, especially if you'd like to reason that your proposed structure also allows for malformed extensions (and that there's a value to that), or that you feel confident that you can implement and maintain all the convenience APIs in related shape.
If that's the case, my only request is that we diverge from what ICU4C is doing with it's ever-growing flat API space of `getUnicodeExtensionKeyword`, `getUnicodeExtensionKeywordAndValue`, `setUnicodeExtensionValue` and so on.
Even if internally you'll shift, I'd like the ergonomic API to look more like:

```rust
locale.extensions.unicode.attributes.iter();
locale.extensions.unicode.attributes.has_key();
locale.extensions.unicode.attributes.value();
locale.extensions.unicode.attributes.set_value();
```

**Comment by sffc** [comment]
Thanks for the detailed reply.

> I don't fully understand why (1) is hard. The compare function has to navigate a tree, rather than iterate a list, but that's it. It's a single tree, with well defined structure and a high-performance comparison will "zip" compare between those two.

You may be right. I'll attempt to scale https://github.com/unicode-org/icu4x/pull/1704 to support extensions with the tree model and see if it works for me.

> As for (2) shouldn't the allowlist be per-extension as well? It seems that a list of values you'd want to preserve shouldn't be "blank-slate" and should narrow values in particular extensions. Is there a use case which doesn't care?

I am trying to implement step 4 in the normalization algorithm in [flexible vertical fallback](https://docs.google.com/document/d/1Mp7EUyl-sFh_HZYgyeVwj88vJGpCBIWxzlCwGgLCDwM/edit#heading=h.8scpxbk3ciej). It says that I want to retain `-u-rg`, `-u-sd`, and one additional extension (which is currently only a Unicode extension), and drop all others. With the tree model, I need to walk the tree to clear everything out. This seems brittle to be done by client code. If the "tree walking" could be done within ICU4X, it would be better, but there is no common interface right now to represent an arbitrary extension.

Basically the operation I was thinking about is something along the lines of

```rust
locale.remove_if(|key| key != "u-rg" && key != "u-sd" && key != special_key)
```

Could you show me how you would implement this operation using the current tree-based API?

**Comment by sffc** [comment]
Discussion with Zibi:

Case 2 could be handled by a "retain" operation.

Zibi suggested:

```rust
locale.extensions.transform = None;
locale.extensions.private = None;
locale.extensions.unicode.remove_if(|(key, _)| !["rg", "sd"].includes(key);
```

I suggested replacing the first two lines with:

```rust
locale.extensions.retain_types(&[ExtensionType::Unicode])
```

Zibi also suggested the following as a more general solution:

```rust
for type in ExtensionType::iter() {
  match type {
    ExtensionType::Transform => loc.ext.transform = None;
    ExtensionType::Private => loc.ext.private = None;
    ExtensionType::Unicode => loc.ext.unicode.keywords.retain(||);
  } 
}
```

**Comment by Manishearth** [comment]
I don't have strong opinions here but I do like the direction this discussion is going. (I wouldn't block on my review here, I think between the two of you all of the use cases I care about are adequately cared for)
